experiment/oodatamodel/oodatamodel.go

```diff
@@ -0,0 +1,156 @@
+				Success: connect.Error == nil,
+			},
+		})
+	}
+	return out
+}
+
+func makeFailure(err error) (s *string) {
+	if err != nil {
+		serio := err.Error()
+		s = &serio
+	}
+	return
+}
+
+// HTTPTor contains Tor information
+type HTTPTor struct {
+	ExitIP   *string `json:"exit_ip"`
+	ExitName *string `json:"exit_name"`
+	IsTor    bool    `json:"is_tor"`
+}
+
+// HTTPBody is an HTTP body. We use this helper class to define a custom
+// JSON encoder that allows us to choose the proper representation depending
+// on whether the Value field is UTF-8 or not.
+type HTTPBody struct {
+	Value string
+}
+
+// MarshalJSON marshal the body to JSON following the OONI spec that says
+// that UTF-8 bodies are represened as string and non-UTF-8 bodies are
+// instead represented as `{"format":"base64","data":"..."}`.
+func (hb HTTPBody) MarshalJSON() ([]byte, error) {
+	if utf8.ValidString(hb.Value) {
+		return json.Marshal(hb.Value)
+	}
+	er := make(map[string]string)
+	er["format"] = "base64"
+	er["data"] = base64.StdEncoding.EncodeToString([]byte(hb.Value))
+	return json.Marshal(er)
+}
+
+// HTTPRequest contains an HTTP request
+type HTTPRequest struct {
+	Body        HTTPBody          `json:"body"`
+	Headers     map[string]string `json:"headers"`
+	Method      string            `json:"method"`
+	Tor         HTTPTor           `json:"tor"`
+	URL         string            `json:"url"`
+	XBodyIsSnap bool              `json:"x_body_is_snap"`
+}
+
+// HTTPResponse contains an HTTP response
+type HTTPResponse struct {
+	Body        HTTPBody          `json:"body"`
+	Code        int64             `json:"code"`
+	Headers     map[string]string `json:"headers"`
+	XBodyIsSnap bool              `json:"x_body_is_snap"`
+}
+
+// RequestEntry is one of the entries that are part of
+// the "requests" key of a OONI report.
+type RequestEntry struct {
+	Failure  *string      `json:"failure"`
+	Request  HTTPRequest  `json:"request"`
+	Response HTTPResponse `json:"response"`
+}
+
+// RequestList is a list of RequestEntry
+type RequestList []RequestEntry
+
+// NewRequestList returns the list for "requests"
+func NewRequestList(httpresults *porcelain.HTTPDoResults) RequestList {
+	// TODO(bassosimone): here I'm using netx snapshots which are
+	// limited to 1<<20. They are probably good enough for a really
+	// wide range of cases, and truncating the body seems good for
+	// loading measurements on mobile as well. I think I should make
+	// sure I modify the documentation to mention that.
+	var out RequestList
+	if httpresults == nil {
+		return out
+	}
+	in := httpresults.TestKeys.HTTPRequests
+	// OONI's data format wants more recent request first
+	for idx := len(in) - 1; idx >= 0; idx-- {
+		var entry RequestEntry
+		entry.Failure = makeFailure(in[idx].Error)
+		entry.Request.Headers = make(map[string]string)
+		for key, values := range in[idx].RequestHeaders {
+			for _, value := range values {
+				entry.Request.Headers[key] = value
+				break // known data format issue
```

Comment:
It may useful to further elaborate on this comment and mention that we are ignoring the second header with the same name.

Reply:
Sure. Just to be clear, this is because our data model uses a map.

Reply:
Code updated to be much more clear.